File: Patterns/Manuscript/manuscript/code/java/DesignPatterns/src/main/java/singleton/Library.java
package singleton;

public class Library {
    private static Library uniqueInstance;
    private Library() {}
    public static synchronized Library getInstance() {
        if (uniqueInstance == null) {
            uniqueInstance = new Library();
        }
        System.out.println("Granting access to the central library");
        return uniqueInstance;
    }

    public void borrowABook() {
        System.out.println("Borrowing a book from the library...");
    }
}

File: code/csharp/DesignPatterns/singleton/Library.cs
namespace Singleton;
using System;

sealed class Library
{
    private static Library _uniqueInstance;
    private static readonly object _lock = new object();

    private Library() { }

    public static Library GetInstance()
    {
        lock (_lock)
        {
            if (_uniqueInstance == null)
            {
                _uniqueInstance = new Library();
                Console.WriteLine("Granting access to the central library");
            }
        }
        return _uniqueInstance;
    }

    public void BorrowABook()
    {
        Console.WriteLine("Borrowing a book from the library...");
    }
}
